services/discovery/src/app.py

```diff
@@ -0,0 +1,227 @@
+    local_buckets = {x['Name'] for x in get_in(['coeffects', 's3', 'Buckets'], world, [])}
+    output = {
+        'IsAuditAccount': trail_bucket in local_buckets,
+        'AuditCloudTrailBucketName': trail_bucket,
+    }
+    return update_in(world, ['output'], lambda x: merge(x or {}, output))
+
+
+def connected(world):
+    output = {
+        'IsConnectedAccount': True,
+    }
+    return update_in(world, ['output'], lambda x: merge(x or {}, output))
+
+
+def safe_check(schema, data):
+    try:
+        return schema(data)
+    except Exception:
+        return None
+
+
+def keep_valid(schema, xs):
+    return [
+        y for y in [safe_check(schema, x) for x in xs]
+        if y is not None
+    ]
+
+
+MINIMUM_CLOUDTRAIL = Schema({
+    "S3BucketName": str,
+    "SnsTopicName": str,
+    "SnsTopicARN": str,
+    "IsMultiRegionTrail": True,
+    "TrailARN": str,
+}, extra=ALLOW_EXTRA, required=True)
+
+
+IDEAL_CLOUDTRAIL = MINIMUM_CLOUDTRAIL.extend({
+    "IsOrganizationTrail": True,
+}, extra=ALLOW_EXTRA, required=True)
+
+
+def cloudtrail(world):
+    trails = get_in(['coeffects', 'cloudtrail', 'trailList'], world, [])
+    valid_trails = keep_valid(IDEAL_CLOUDTRAIL, trails) or keep_valid(MINIMUM_CLOUDTRAIL, trails)
+    trail_topic = valid_trails[0]['SnsTopicName'] if valid_trails else None
+    account_id = trail_topic.split(':')[4] if trail_topic else None
+    output = {
+        'IsCloudTrailAccount': account_id == event_account_id(world),
+        'CloudTrailSNSTopicName': trail_topic,
+    }
+    return update_in(world, ['output'], lambda x: merge(x or {}, output))
+
+
+MINIMUM_BILLING_REPORT = Schema({
+    'TimeUnit': 'HOURLY',
+    'Format': 'textORcsv',
+    'Compression': 'GZIP',
+    'AdditionalSchemaElements': ExactSequence(['RESOURCES']),
+    'S3Bucket': str,
+    'S3Prefix': str,
+    'S3Region': str,
+    'ReportVersioning': 'CREATE_NEW_REPORT',
+    'RefreshClosedReports': True,
+}, extra=ALLOW_EXTRA, required=True)
+
+
+def master_payer(world):
+    report_definitions = get_in(['coeffects', 'cur', 'ReportDefinitions'], world, [])
+    local_buckets = {x['Name'] for x in get_in(['coeffects', 's3', 'Buckets'], world, [])}
+    valid_report_definitions = keep_valid(MINIMUM_BILLING_REPORT, report_definitions)
+    default = valid_report_definitions[0]['S3Bucket'] if any(valid_report_definitions) else None
+    valid_local_report_definitions = [x for x in valid_report_definitions if x['S3Bucket'] in local_buckets]
+    local = any(valid_local_report_definitions)
+    output = {
+        'IsMasterPayerAccount': local,
+        'MasterPayerBillingBucketName': valid_local_report_definitions[0]['S3Bucket'] if local else default,
+    }
+    return update_in(world, ['output'], lambda x: merge(x or {}, output))
+
+
+def discover_account_types(world):
+    return pipe(world,
+                audit,
+                connected,
+                cloudtrail,
+                master_payer)
```

Comment:
Still kinda weirded out by functions that are nouns...

Reply:
Done